galaxyproject/galaxy
# Replace sqlalchemy-utilities with local implementation

## What did you do? 
Replaced dependency on sqlalchemy_utils with a local implementation of the functionality we actually need.

## Why did you make this change?
The main goal of this is to get this dependency out of the way and move on with migrating Galaxy to SA 1.4. It is designed as a temporary or permanent replacement for sqlalchemy-utils for reasons outlined in #11459. This may be replaced with sqlalchemy-utils once that library supports SA 1.4 (effort is underway).

To do: (lower priority; after this is merged)
1. Postgresql and MySQL tests have been run locally (they passed); however they will be skipped within the unit test workflow, since it doens't include postgres and mysql services. Since these (and `unit/data/test_galaxy_mapping`) are the only unit tests that require a database, I'm not sure what's the best way to handle this: add the serivces or setup a separate workflow.

2. Handling of views can be improved: we have setup a DIY mechanism for installing them in the database; however, SQLAlchemy provides a simple way to do it via `after_create`/`before_drop` hooks: we would just need to use `event.listen(metadata, 'after_create', [our DDLElement])`. This is the approach taken by sqlalchemy-utils; however, we don't use it and pass an empty metadata object instead ([see comment](https://github.com/ic4f/galaxy/blob/dev_no_sqlalchemy_utils/lib/galaxy/model/view/utils.py#L34)). I did not modify this as this to keep the PR focused on one thing.

Implementation Notes

1. I considered using mocks for tests (sqlalchemy-utils does that), but decided that for testing the creation of a database or a database object such as a view it makes sense to use a real database for increased test fidelity (because this is outside the context of what SA already does and what's covered by SA's internal tests). So, technically, this makes them integration tests. However, in the context of Galaxy's testing frameworks integration tests run Galaxy in different configurations, which is not what these tests are. So I left them under unit tests (they are also fast, so that's fitting).

2. I've dropped the things we don't use (creating materialized views, cascade on view drop, using after_create/before_drop hooks). Any of this can be added when needed. Until then, yagni.

3. I've changed the singature for `create_database` and `database_exists` to 2 parameters: the connection URL and an optional database to manipulate. This addresses the issue of some installations not having access to the default `postgres` database and is similar to the solution proposed [here](https://github.com/kvesteri/sqlalchemy-utils/pull/463#issuecomment-668214945).

3. This code is not SA 1.4-compatible. It was first implemented as such, but other parts of Galaxy broke under SA 1.4, so this was downgraded to 1.3. It will be migrated to run under 1.4 with the rest of Galaxy.


## How to test the changes? 
Manually: 
- Your `database_connection` should be unset. Delete local sqlite database, run Galaxy; database should be created.
- Your `database_connection` should be using postgresql; point it to a nonexistant database
- To run the new unit tests: 
  `GALAXY_TEST_CONNECT_POSTGRES_URI='postgresql://postgres@localhost:5432/postgres' pytest test/unit/model/`
- Repeat the postgresql steps for mysql: 
  `GALAXY_TEST_CONNECT_MYSQL_URI='mysql+mysqldb://root@localhost/mysql' pytest test/unit/model/`
  (you'll need to `python -m pip install mysqlclient`; you'll also need a running instance of mysql)
